runtime/connectors/blob/blobdownloader.go

```diff
@@ -3,119 +3,272 @@ package blob
-		// Access storage.Query via q here.
-		var q *storage.Query
-		if as(&q) {
-			// we only need name and size, adding only required attributes to reduce data fetched
-			_ = q.SetAttrSelection([]string{"Name", "Size"})
-		}
-		return nil
-	}}
+	return &blobIterator{
+		ctx:     ctx,
+		bucket:  bucket,
+		objects: items,
+		tempDir: tempDir,
+	}, nil
+}
 
-	if prefix != "." {
-		listOptions.Prefix = prefix
+func (iter *blobIterator) Close() error {
+	fileutil.ForceRemoveFiles(iter.localFiles)
+	// remove temp dir since recursive paths created have to be removed as well
+	os.RemoveAll(iter.tempDir)
+	return iter.bucket.Close()
+}
+
+func (iter *blobIterator) HasNext() bool {
+	return iter.index < len(iter.objects)
+}
+
+// NextBatch downloads next n files and copies to local directory
+// Callers responsibility to delete files once done
+// Thread unsafe
+func (iter *blobIterator) NextBatch(n int) ([]string, error) {
+	if !iter.HasNext() {
+		return nil, io.EOF
 	}
 
-	var size, fetched int64
-	var matchCount int
-	var fileNames []string
+	// delete files created in last iteration
+	fileutil.ForceRemoveFiles(iter.localFiles)
+	start := iter.index
+	end := iter.index + n
+	if end > len(iter.objects) {
+		end = len(iter.objects)
+	}
+	iter.index = end
+
+	// this is not necessary on every iteration
+	// but there may be cases where n in first batch is different from n in next batch
+	// to keep things easy creating a new slice every time
+	iter.localFiles = make([]string, n)
+	g, grpCtx := errgroup.WithContext(iter.ctx)
+	g.SetLimit(_concurrentBlobDownloadLimit)
+	for i, obj := range iter.objects[start:end] {
+		obj := obj
+		index := start + i // with repect to object slice
+		g.Go(func() error {
+			// need to create file by maintaining same dir path as in glob for hivepartition support
+			dir := filepath.Join(iter.tempDir, filepath.Dir(obj.obj.Key))
+			// filename
+			objName := filepath.Base(obj.obj.Key)
+			file, err := fileutil.OpenTempFileInDir(dir, objName)
+			if err != nil {
+				return err
+			}
+			defer file.Close()
+
+			iter.localFiles[index-start] = file.Name()
+			ext := fileutil.FullExt(obj.obj.Key)
+			partialReader, isPartialDownloadSupported := _partialDownloadReaders[ext]
+			if obj.full || !isPartialDownloadSupported {
+				// download full file
+				return downloadObject(grpCtx, iter.bucket, obj.obj.Key, file)
+			}
+			// download partial file
+			// check if, for smaller size we can download entire file
+			switch partialReader {
+			case "parquet":
+				return downloadParquet(grpCtx, iter.bucket, obj.obj, obj.extractOption, file)
+			case "csv":
+				return downloadCSV(grpCtx, iter.bucket, obj.obj, obj.extractOption, file)
+			default:
+				// should not reach here
+				panic(fmt.Errorf("partial download not supported for extension %q", ext))
+			}
+		})
+	}
+
+	if err := g.Wait(); err != nil {
+		return nil, err
+	}
+
+	result := make([]string, end-start)
+	copy(result, iter.localFiles)
+	return result, nil
```

Comment:
This is to stop clients from making any modification to the slice of files. Since go doesn't have immutable slice so creating a new slice. Found this in the style guide.

Reply:
Ahh, got it – because if they mess up `iter.localFiles`, then we won't correctly delete the temp files on next iteration? Would be good to put a small comment about that here